MM 50960 - Add company name and invite members pages to preparing-workspase for self-hosted

pvev opened the pull request:
#### Summary
This PR adds the company name form and the invite members screens to the preparing workspace intro for first-admins. The company name is used to create the first team.

#### Ticket Link
https://mattermost.atlassian.net/browse/MM-50960

#### Screenshots

https://user-images.githubusercontent.com/10082627/230107287-52c2705b-6d5a-4b13-a8fd-27555ff940d2.mov



#### Release Note
```release-note
Add a form page to capture company name data and the invite members section to the preparing-workspace initial screen for self-hosted installations
```

mattermost-build commented:
E2E tests not automatically triggered, because PR has no approval yet. Please ask a developer to review and then try again to attach the QA label.

neallred reviewed (changes requested):
LGTM overall! I think we want to prefix the svg's ids, and unless there's some product owner reason to not show invite members to cloud instances, I think we should show it to cloud instances too.

pvev commented on webapp/channels/src/components/preparing_workspace/preparing_workspace.tsx:
```diff
@@ -98,28 +107,35 @@ const PreparingWorkspace = (props: Props) => {
     const stepOrder = [
+        isSelfHosted && WizardSteps.Organization,
         pluginsEnabled && WizardSteps.Plugins,
+        isSelfHosted && WizardSteps.InviteMembers,
```
there is a change coming to cloud instances, will be a little bit different from self-hosted https://mattermost.atlassian.net/browse/MM-50013

neallred commented on webapp/channels/src/components/preparing_workspace/preparing_workspace.tsx:
```diff
@@ -98,28 +107,35 @@ const PreparingWorkspace = (props: Props) => {
     ].filter((x) => Boolean(x)) as WizardStep[];
 
+    // first steporder that is not false
+    const firstShowablePage = stepOrder.filter((x) => Boolean(x))[0];
```
Forgot to mention; this can just be `const firstShowablePage = stepOrder[0]` since the definition above already filters out all the false ones.

pvev commented on webapp/channels/src/components/preparing_workspace/invite_members_link.tsx:
```diff
@@ -0,0 +1,64 @@
+
+const InviteMembersLink = (props: Props) => {
+    const copyText = useCopyText({
+        trackCallback: () => trackEvent('FIXME', 'admin_setup_click_copy_invite_link'),
```
@esethna do you find any value on tracking when a user clicks the copy invite link? if so, what category do you think we should use for this?
cc: @marianunez

annelieseklein replied:
yes i think we should track that absolutely @pvev

marianunez replied:
For the category we should be consistent with which ever category we had until now for onboarding. However, also noticed in this PR we are using the `first_admin_setup` category. Any preference @esethna ?

pvev commented:
Fixed the invite link not containing the teamId issue. Once the organization name is entered and the next button is clicked, the team is created. If for some reason the organization name is changed the team display name is also modified.
Also covers the scenario if the next button is clicked (team created) and then the page is reloaded, it will find there is already a team and will only update the team display name instead of creating a new team.

@JulienTant @marianunez @esethna ^^


https://user-images.githubusercontent.com/10082627/231233304-d93b3369-9b6e-4cf6-91e9-8e2f0c4daa3a.mov

annelieseklein commented:
LGTM overall! thanks for the very clear video with clicking all the states! is there a weird movement on the integrations screen when you click continue? it looks like "skip" moves to the right a bit when you click

pvev commented:
> LGTM overall! thanks for the very clear video with clicking all the states! is there a weird movement on the integrations screen when you click continue? it looks like "skip" moves to the right a bit when you click

yes, with my latest commit I fixed that the continue button was getting a little bit wider during the on focus event


https://user-images.githubusercontent.com/10082627/231452070-2f9da936-e456-4ad7-96ab-94da60eb5ab7.mov

neallred reviewed (changes requested):
Thanks for the added unit tests! Can we migrate them to testing-library to support web platform team's effort to migrate all enzyme tests to testing-library?

Everything else about the PR looks good to me :+1: .

marianunez reviewed (commented):
@pvev I haven't done the code review yet but noticed that in cloud the onboarding hangs in the final screen of `Preparing Workspace`. Refreshing starts the onboarding again. You can try it in the current server that is setup

https://user-images.githubusercontent.com/936315/231613488-823be559-f3c7-4943-9748-14d0d447a337.mov

marianunez commented on webapp/channels/src/components/preparing_workspace/invite_members.tsx:
```diff
@@ -0,0 +1,114 @@
+
+    const inviteURL = useMemo(() => {
+        let urlBase = '';
+        if (props.configSiteUrl && !props.configSiteUrl.includes('localhost')) {
+            urlBase = props.configSiteUrl;
+        } else if (props.formUrl && !props.formUrl.includes('localhost')) {
+            urlBase = props.formUrl;
+        } else {
+            urlBase = props.browserSiteUrl;
+        }
+        return `${urlBase}/signup_user_complete/?id=${props.teamInviteId}`;
```
What is the difference between `browserSiteUrl` and `props.configSiteUrl` ?

It is most likely the SiteUrl has not been set yet so not sure how effective this invite screen is so early in the process

pvev replied:
yeah, probably no need to use the configsiteurl.

marianunez replied:
Yes, and what is `browserSiteUrl`? Same as `formUrl`? Just confused why we have 3

marianunez commented on webapp/channels/src/components/preparing_workspace/organization.tsx:
```diff
@@ -0,0 +1,207 @@
+        const thereIsAlreadyATeam = teams.length > 0;
+        teamApiError.current = null;
+
+        if (!validation.error && !thereIsAlreadyATeam && props.isSelfHosted) {
+            createTeamFromOrgName();
+        } else if (!validation.error && thereIsAlreadyATeam && props.isSelfHosted) {
+            updateTeamNameFromOrgName();
+        }
```
These checks for `isSelfHosted` doesn't hurt but I believe this screen is not even shown if not selfhosted

pvev replied:
you are right, these can be removed.

marianunez commented on webapp/channels/src/components/preparing_workspace/organization.tsx:
```diff
@@ -0,0 +1,207 @@
+                        <div className='Organization__progress-path'>
+                            <OrganizationSVG/>
+                            <PageLine
+                                style={{
+                                    marginTop: '5px',
+                                    height: 'calc(50vh)',
+                                }}
```
Should this be moved into the scss?

pvev replied:
since the PageLine component would need these values to be different depending on where it is used, this act more as config params than as css values. Moving it to css would mean having to have different options for each scenario and passing to PageLine component a param which will be used to determine which css style classname with what marginTop and heigh to use.

marianunez replied:
Thanks for that context!

marianunez reviewed (changes requested):
Thanks @pvev & @JulienTant! Some minor notes below but overall blocker is that doesn't seem to complete in Cloud as noted in the post above

neallred reviewed (approved):
LGTM, thanks @pvev

pvev commented:
the cloud workspace preparation workflow is fixed now @marianunez:

https://user-images.githubusercontent.com/10082627/231830334-2558ff42-2f98-48f6-9c78-d62236666f54.mov

marianunez reviewed (approved):
Thanks @pvev! LGTM 🎉

stevemudie commented:
/e2e-test

marianunez commented:
/branch-update

marianunez commented:
> Mattermost test server created! 🎉
> 
> Access here: https://mattermost-server-pr-22838.test.mattermost.cloud
> 
> Account Type	Username	Password
> Admin	sysadmin	Sys@dmin123
> User	user-1	User-1@123

@stevemudie finally got the server up, if you want to take a look

stevemudie commented:
Thanks @marianunez! It looks good and worked as expected.

I found it interesting that I could use the copied invite link in a new tab to directly open up the Workspace. And that I could go back, edit and change the company name which also changed the team name. But no issues as that's all expected I'm guessing.

marianunez commented:
> Thanks @marianunez! It looks good and worked as expected.
> 
> I found it interesting that I could use the copied invite link in a new tab to directly open up the Workspace. And that I could go back, edit and change the company name which also changed the team name. But no issues as that's all expected I'm guessing.

You would probably need to use the invite link in incognito in the same PC as it was in session for that admin. 

Also, can we smoke test other users (not first admin) not getting this experience and also cloud to make sure there are no regressions?

JulienTant reviewed (approved):
Awesome @pvev thank you!!! The magic of the animations 👍🏻 so nice

marianunez commented:
@annelieseklein @esethna gentle ping on this review so we can make it for v7.11

annelieseklein commented:
looking good! 

couple small state issues:
<img width="211" alt="Screen Shot 2023-04-17 at 4 10 36 PM" src="https://user-images.githubusercontent.com/29821919/232629472-ddb5293c-e6b5-450c-a90a-b4f0c3a88c54.png">
1. "Skip" shouldn't have a underline here. 

<img width="194" alt="Screen Shot 2023-04-17 at 4 12 10 PM" src="https://user-images.githubusercontent.com/29821919/232629871-1991fb57-7043-4c48-9fee-b702c633b25d.png">

2. it looks like the copy link button is missing an active state (after the hover state)

you can take a look at all button states here in this file: https://www.figma.com/file/2uYWxjnMJ9IQDOba9b9bfV/Components---Button?node-id=3558%3A13159&t=IOEK7alGUcog4zPw-1

esethna commented:
I couldn't test as the team was already set up, but looks good based on the screen videos in the PR description, I'll defer to @annelieseklein for any detailed UI feedback. Approving to unblock merge

amyblais commented:
/cherry-pick cloud

mattermost-build commented:
Cherry pick is scheduled.

mattermost-build commented:
Error trying doing the automated Cherry picking. Please do this manually

```
+++ Updating remotes...
Fetching upstream
Failed to add the RSA host key for IP address '140.82.113.3' to the list of known hosts (/app/.ssh/known_hosts).
From github.com:mattermost/mattermost-server
   a24111f9bd..1f57819054  master                -> upstream/master
   f0ed400732..3ca789979a  MM-50966-in-product-expansion-backend -> upstream/MM-50966-in-product-expansion-backend
   61bbc442e6..cb0f4a196c  cloud                 -> upstream/cloud
 * [new branch]            fix-boards-postgres-schema-migration-needed-check -> upstream/fix-boards-postgres-schema-migration-needed-check
   ad494a961f..f4f234a09c  github-plugin-2.1.5   -> upstream/github-plugin-2.1.5
   ef95c1f88f..ec88267622  manual-cherry-pick-pr-22917 -> upstream/manual-cherry-pick-pr-22917
   a8c02e6414..0e0f86b1b8  manual-cherry-pick-pr-22917-3 -> upstream/manual-cherry-pick-pr-22917-3
   e0c0c24f8d..865af40592  persist-notifications -> upstream/persist-notifications
   e35813d3e8..9b665c68ac  release-7.1           -> upstream/release-7.1
   1ada1a5379..8985f18cf2  release-7.10          -> upstream/release-7.10
   55ce33cf74..e05ee18add  release-7.8           -> upstream/release-7.8
   012cfa4e83..4108f4bfc5  release-7.9           -> upstream/release-7.9
Fetching origin
Failed to add the RSA host key for IP address '140.82.113.3' to the list of known hosts (/app/.ssh/known_hosts).
+++ Updating remotes done...
+++ Creating local branch automated-cherry-pick-of-mattermost-server-#22838-upstream-cloud-1681914257
Switched to a new branch 'automated-cherry-pick-of-mattermost-server-#22838-upstream-cloud-1681914257'
Branch 'automated-cherry-pick-of-mattermost-server-#22838-upstream-cloud-1681914257' set up to track remote branch 'cloud' from 'upstream'.

+++ About to attempt cherry pick of PR #22838 with merge commit b5c48e75b385533cd3dc329b658cb47768a7b992.

error: could not apply b5c48e75b3... MM 50960 - Add company name and invite members pages to preparing-workspace for self-hosted (#22838)
hint: after resolving the conflicts, mark the corrected paths
hint: with 'git add <paths>' or 'git rm <paths>'
hint: and commit the result with 'git commit'

+++ Conflicts detected:

UU e2e-tests/playwright/support/server/default_config.ts
Aborting.

+++ Aborting in-progress git cherry-pick.

+++ Returning you to the master branch and cleaning up.

```

amyblais commented:
Needs a manual cherry-pick to ``cloud``.

amyblais commented:
/cherry-pick cloud-latest

mattermost-build commented:
Cherry pick is scheduled.

mattermost-build commented:
Error trying doing the automated Cherry picking. Please do this manually

```
+++ Updating remotes...
Fetching upstream
Failed to add the RSA host key for IP address '140.82.114.4' to the list of known hosts (/app/.ssh/known_hosts).
Fetching origin
Failed to add the RSA host key for IP address '140.82.114.4' to the list of known hosts (/app/.ssh/known_hosts).
+++ Updating remotes done...
+++ Creating local branch automated-cherry-pick-of-mattermost-server-#22838-upstream-cloud-latest-1682450064
Switched to a new branch 'automated-cherry-pick-of-mattermost-server-#22838-upstream-cloud-latest-1682450064'
Branch 'automated-cherry-pick-of-mattermost-server-#22838-upstream-cloud-latest-1682450064' set up to track remote branch 'cloud-latest' from 'upstream'.

+++ About to attempt cherry pick of PR #22838 with merge commit b5c48e75b385533cd3dc329b658cb47768a7b992.

error: could not apply b5c48e75b3... MM 50960 - Add company name and invite members pages to preparing-workspace for self-hosted (#22838)
hint: after resolving the conflicts, mark the corrected paths
hint: with 'git add <paths>' or 'git rm <paths>'
hint: and commit the result with 'git commit'

+++ Conflicts detected:

UU e2e-tests/playwright/support/server/default_config.ts
UU webapp/channels/src/utils/constants.tsx
Aborting.

+++ Aborting in-progress git cherry-pick.

+++ Returning you to the master branch and cleaning up.

```

amyblais commented:
/cherry-pick cloud

mattermost-build commented:
Cherry pick is scheduled.
